### PR TITLE
Add support Influxdb basic auth

### LINKED ISSUE #7644: Add support for passing username and password to InfluxDB using basic auth
## Is your feature request related to a problem? Please describe.

We're trying out https://github.com/VictoriaMetrics/VictoriaMetrics as a replacement for InfluxDB. The problem that it supports only basic auth and doesn't work when user/pass pair is passed as arguments.

## Describe the solution you'd like

An option allowing to change behavior would be nice.

## Describe alternatives you've considered

N/A

## Additional context

ref https://github.com/VictoriaMetrics/VictoriaMetrics/issues/240

### PR DESCRIPTION
**Pull request by Yonas-net**
fixes #7644

### CODE REVIEW

File: lib/perfdata/influxdbwriter.cpp [hunk 1]
```diff
@@ -504,6 +505,16 @@ void InfluxdbWriter::Flush()
 	request.set(http::field::user_agent, "Icinga/" + Application::GetAppVersion());
 	request.set(http::field::host, url->GetHost() + ":" + url->GetPort());
 
+	Dictionary::Ptr basic_auth = GetBasicAuth();
```
Comment (Al2Klimov) [inline lib/perfdata/influxdbwriter.cpp:508]: Please create new variables with camelCase.

File: lib/perfdata/influxdbwriter.cpp [hunk 2]
```diff
@@ -504,6 +505,16 @@ void InfluxdbWriter::Flush()
...
+
+	if (basic_auth) {
+		basic_auth->Set("username", basic_auth->Get("username"));
+		basic_auth->Set("password", basic_auth->Get("password"));
+		request.set(http::field::authorization, "Basic " + Base64::Encode(basic_auth->Get("username") + ":" + basic_auth->Get("password")));
+
+		Log(LogInformation, "Icinga2 uses basic authentication to connect to InfluxDB server");
```
Comment (Al2Klimov) [inline lib/perfdata/influxdbwriter.cpp:513]: What's this for?

File: lib/perfdata/influxdbwriter.cpp [hunk 3]
```diff
@@ -603,3 +614,14 @@ void InfluxdbWriter::ValidateServiceTemplate(const Lazy<Dictionary::Ptr>& lvalue
...
+void InfluxdbWriter::ValidateBasicAuth(const Lazy<Dictionary::Ptr>& lvalue, const ValidationUtils& utils) {
+	ObjectImpl<InfluxdbWriter>::ValidateBasicAuth(lvalue, utils);
+
+	String username = lvalue()->Get("username");
+	if (username.IsEmpty())
+		BOOST_THROW_EXCEPTION(ValidationError(this, { "basic_auth", "username" }, "Basic authentication username must to be not empty."));
+
+	String password = lvalue()->Get("password");
+	if (password.IsEmpty())
+		BOOST_THROW_EXCEPTION(ValidationError(this, { "basic_auth", "password" }, "Basic authentication password must to be not empty."));
```
Comment (Al2Klimov) [inline lib/perfdata/influxdbwriter.cpp:624]: Do you actually need this?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: lib/perfdata/influxdbwriter.cpp [hunk 4]
```diff
@@ -504,6 +505,12 @@ void InfluxdbWriter::Flush()
 	request.set(http::field::user_agent, "Icinga/" + Application::GetAppVersion());
 	request.set(http::field::host, url->GetHost() + ":" + url->GetPort());
 
+	Dictionary::Ptr basicAuth = GetBasicAuth();
+
+	if (basicAuth) {
+		request.set(http::field::authorization, "Basic " + Base64::Encode(basicAuth->Get("username") + ":" + basicAuth->Get("password")));
+	}
```
Comment (Al2Klimov) [inline lib/perfdata/influxdbwriter.cpp:512]: I'd make this a `{` `}` block.

File: lib/perfdata/influxdbwriter.cpp [hunk 5]
```diff
@@ -603,3 +610,6 @@ void InfluxdbWriter::ValidateServiceTemplate(const Lazy<Dictionary::Ptr>& lvalue
 	}
 }
 
+void InfluxdbWriter::ValidateBasicAuth(const Lazy<Dictionary::Ptr>& lvalue, const ValidationUtils& utils) {
+	ObjectImpl<InfluxdbWriter>::ValidateBasicAuth(lvalue, utils);
+}
```
Comment (Al2Klimov) [inline lib/perfdata/influxdbwriter.cpp:615]: What's this for?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: lib/perfdata/influxdbwriter.cpp [hunk 6]
```diff
@@ -602,4 +614,3 @@ void InfluxdbWriter::ValidateServiceTemplate(const Lazy<Dictionary::Ptr>& lvalue
 		}
 	}
 }
-
```
Comment (Al2Klimov) [inline lib/perfdata/influxdbwriter.cpp:605]: Not part of the task.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Write a test protocol.

**Comment by Yonas-net** [comment]
Testprotocol: Basic_Auth

With the configs shown above icinga2 output looks like this.

```
icinga2_1           | [2020-10-07 15:14:33 +0000] warning/InfluxdbWriter: Unexpected response code: Unauthorized
nginx-service_1     | 172.18.0.2 - - [07/Oct/2020:15:14:33 +0000] "POST /write?db=icinga2&precision=s HTTP/1.0" 401 179 "-" "Icinga/v2.12.0-3-gdf8e0cd8a" "-"
icinga2_1           | [2020-10-07 15:14:33 +0000] warning/InfluxdbWriter: Unexpected Content-Type: text/html
icinga2_1           | [2020-10-07 15:14:43 +0000] warning/InfluxdbWriter: Unexpected response code: Unauthorized
icinga2_1           | [2020-10-07 15:14:43 +0000] warning/InfluxdbWriter: Unexpected Content-Type: text/html
nginx-service_1     | 172.18.0.2 - - [07/Oct/2020:15:14:43 +0000] "POST /write?db=icinga2&precision=s HTTP/1.0" 401 179 "-" "Icinga/v2.12.0-3-gdf8e0cd8a" "-"
```

And then after I add the correct authentication config to the ``Icinga2 influxdb.conf`` file.
```
basic_auth = {
     username = "icinga"
     password = "icinga"
}
```

Icinga2 is able to connect to InluxDB server using the basic auth method.
```
icinga2_1           | [2020-10-07 15:20:04 +0000] information/WorkQueue: #5 (InfluxdbWriter, influxdb) items: 0, rate: 0.0166667/s (1/min 1/5min 1/15min);
influxdb-service_1  | [httpd] 172.20.0.4 - icinga [07/Oct/2020:15:20:14 +0000] "POST /write?db=icinga2&precision=s HTTP/1.0" 204 0 "-" "Icinga/v2.12.0-3-gdf8e0cd8a" 9979f4dc-08b0-11eb-8001-0242ac140003 17204
nginx-service_1     | 172.20.0.2 - icinga [07/Oct/2020:15:20:14 +0000] "POST /write?db=icinga2&precision=s HTTP/1.0" 204 0 "-" "Icinga/v2.12.0-3-gdf8e0cd8a" "-"
influxdb-service_1  | [httpd] 172.20.0.4 - icinga [07/Oct/2020:15:20:24 +0000] "POST /write?db=icinga2&precision=s HTTP/1.0" 204 0 "-" "Icinga/v2.12.0-3-gdf8e0cd8a" 9f6d9696-08b0-11eb-8002-0242ac140003 2342
nginx-service_1     | 172.20.0.2 - icinga [07/Oct/2020:15:20:24 +0000] "POST /write?db=icinga2&precision=s HTTP/1.0" 204 0 "-" "Icinga/v2.12.0-3-gdf8e0cd8a" "-"
influxdb-service_1  | [httpd] 172.20.0.4 - icinga [07/Oct/2020:15:20:34 +0000] "POST /write?db=icinga2&precision=s HTTP/1.0" 204 0 "-" "Icinga/v2.12.0-3-gdf8e0cd8a" a5647d2d-08b0-11eb-8003-0242ac140003 2912
nginx-service_1     | 172.20.0.2 - icinga [07/Oct/2020:15:20:34 +0000] "POST /write?db=icinga2&precision=s HTTP/1.0" 204 0 "-" "Icinga/v2.12.0-3-gdf8e0cd8a" "-"
```

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Adjust the docs.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Squash everything.

**Review by Al2Klimov — APPROVED** [review]

**Comment by Al2Klimov** [comment]
What happens if you use this branch, but not specify basic auth?

**Comment by Yonas-net** [comment]
> What happens if you use this branch, but not specify basic auth?

I removed basic_auth completely from the config and this happens:
```
nginx-service_1     | 172.20.0.2 - - [08/Oct/2020:11:58:02 +0000] "POST /write?db=icinga2&precision=s HTTP/1.0" 401 179 "-" "Icinga/v2.12.0-3-gdf8e0cd8a" "-"
nginx-service_1     | 172.20.0.2 - - [08/Oct/2020:11:58:12 +0000] "POST /write?db=icinga2&precision=s HTTP/1.0" 401 179 "-" "Icinga/v2.12.0-3-gdf8e0cd8a" "-"
icinga2_1           | [2020-10-08 11:58:12 +0000] warning/InfluxdbWriter: Unexpected response code: Unauthorized
icinga2_1           | [2020-10-08 11:58:12 +0000] warning/InfluxdbWriter: Unexpected Content-Type: text/html
```

**Review by julianbrost — CHANGES_REQUESTED** [review]
With this, the new basic auth credentials can be queried using the API. This is not the case for the existing password field.

```
$ curl -ksSu root:icinga -H 'Accept: application/json' -X GET https://localhost:5665/v1/objects/influxdbwriters/influxdb | jq
[...]
        "basic_auth": {
          "password": "bar",
          "username": "foo"
        },
[...]
```

**Review by julianbrost — APPROVED** [review]
Looks good to me now.

I was also thinking about adding an attribute like `auth_method` instead which can either take `url` or `basic` as a value. This should suffice for what is described in the corresponding issue, however, I guess you can also build setups with a reverse-proxy where you might want to use both authentication mechanisms with different credentials. So fine for me as it is right now.

**Review by Al2Klimov — APPROVED** [review]

**Pull request merged**